Use dirty bits to only traverse the unchanged subtrees with recursiveUpdateLayerPositions.

mattwoodrow opened the pull request:
#### 0ef1fec1cdce7a5dd3007976f09afdf302267593
<pre>
Use dirty bits to only traverse the unchanged subtrees with recursiveUpdateLayerPositions.
<a href="https://bugs.webkit.org/show_bug.cgi?id=276706">https://bugs.webkit.org/show_bug.cgi?id=276706</a>
&lt;<a href="https://rdar.apple.com/131903605">rdar://131903605</a>&gt;

Reviewed by Alan Baradlay.

Adds a set of dirty bits to mark which RenderLayers need their position updates,
and makes recursiveUpdateLayerPositions only walk the dirty subset of the tree.

With a debug build, the full traversal still gets completed, and it verifies
that all computed values would have been a no-op had they been skipped.

recursiveUpdateLayerPositions gets a templated &apos;mode&apos; parameter so that the
compiled code in release should be mostly identical to before (except the flag
checks). Each block of state changes handles the write/verify branches, to keep
these together and hopefully keep them in sync.

Also adds a new tree dump to &apos;Compositing&apos; logging, showing the RenderLayer tree
and its dirty bits before/after the recursiveUpdateLayerPositions.

The general approach is that we mark layers (and their ancestors) that we know
to be changed), and then descend into children of those to check if any ancestor
state that they depend on has changed.

We could instead ensure that changes to state that descendants use trigger a
full descendant subtree update, and avoid needing to iterate the child list to
check. Logging of visited vs updated layer counts suggests that this wouldn&apos;t
make much difference. We would likely also need to make sure these extra
triggers are specific (not all style changes for example) to avoid updating too
much.

Layout on a RenderObject with a layer triggers update on the direct children of
that layer, since we can move (but not layout) a descendant renderer, and the
child layer is then a further descendant of the moved renderer. We need to
update the child layer to account for the different offset to parent layer.

ClipRectsContext always (and only) specified
ClipRectsOption::IncludeOverlayScrollbarSize when using RootRelativeClipRects,
so that is now implicit. Clip rect validation now can assert these when adding
to the cache, and verify cached clips using it.

recursiveUpdateLayerPositionsAfterScroll no longer checks isVisuallyNonEmpty()
to clear repaint rects, since this doesn&apos;t match the condition we use any other
time (and changing the others results in lots of test failures).

* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::repaintOldAndNewPositionsForSVGRenderer const):
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::ClipRectsCache::ClipRectsCache):
(WebCore::ClipRectsCache::getClipRects const):
(WebCore::RenderLayer::RenderLayer):
(WebCore::RenderLayer::addChild):
(WebCore::RenderLayer::isStackingContextChanged):
(WebCore::RenderLayer::setNeedsPositionUpdate):
(WebCore::RenderLayer::setSelfAndChildrenNeedPositionUpdate):
(WebCore::RenderLayer::setSelfAndDescendantsNeedPositionUpdate):
(WebCore::compositingLogEnabledRenderLayer):
(WebCore::RenderLayer::updateLayerPositionsAfterStyleChange):
(WebCore::RenderLayer::updateLayerPositionsAfterLayout):
(WebCore::RenderLayer::ancestorLayerPositionStateChanged):
(WebCore::RenderLayer::recursiveUpdateLayerPositions):
(WebCore::RenderLayer::dirtyAncestorChainHasSelfPaintingLayerDescendantStatus):
(WebCore::RenderLayer::computeRepaintRects):
(WebCore::RenderLayer::computeRepaintRectsIncludingDescendants):
(WebCore::RenderLayer::compositingStatusChanged):
(WebCore::RenderLayer::recursiveUpdateLayerPositionsAfterScroll):
(WebCore::RenderLayer::dirtyAncestorChainHasBlendingDescendants):
(WebCore::RenderLayer::updateTransform):
(WebCore::RenderLayer::setHasVisibleContent):
(WebCore::RenderLayer::dirtyVisibleContentStatus):
(WebCore::RenderLayer::dirtyAncestorChainVisibleDescendantStatus):
(WebCore::RenderLayer::updateLayerPosition):
(WebCore::RenderLayer::hitTestLayer):
(WebCore::RenderLayer::hitTestTransformedLayerInFragments):
(WebCore::RenderLayer::verifyClipRects):
(WebCore::RenderLayer::verifyClipRect):
(WebCore::RenderLayer::updateClipRects):
(WebCore::RenderLayer::clipRects const):
(WebCore::RenderLayer::calculateClipRects const):
(WebCore::outputLayerPositionTreeLegend):
(WebCore::outputLayerPositionTreeRecursive):
(WebCore::showLayerPositionTree):
* Source/WebCore/rendering/RenderLayer.h:
(WebCore::RenderLayer::setIsHiddenByOverflowTruncation):
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::setBackingSharingLayers):
(WebCore::RenderLayerBacking::setRequiresOwnBackingStore):
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::BackingSharingState::issuePendingRepaints):
(WebCore::RenderLayerCompositor::updateBacking):
* Source/WebCore/rendering/RenderLayerScrollableArea.cpp:
(WebCore::RenderLayerScrollableArea::positionOverflowControls):
* Source/WebCore/rendering/RenderLayerScrollableArea.h:
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::RenderObject::clearNeedsLayout):
(WebCore::operator&lt;&lt;):
* Source/WebCore/rendering/RenderObject.h:

Canonical link: <a href="https://commits.webkit.org/284780@main">https://commits.webkit.org/284780@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/edc8f1d9900ba87e4b16c5fcb806b7593efb22a8

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/70522 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/49928 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/23287 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/74611 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/21700 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/72639 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/57728 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/21540 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/74611 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/21700 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/73588 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/45408 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/60795 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/74611 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/42067 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/18230 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/20061 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/63643 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/64000 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/18585 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/76330 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/69769 "Built successfully and passed tests") | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/87/builds/14748 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/17810 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/63602 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/86/builds/14791 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/60862 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/63537 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-1-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/5219 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/91552 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10801 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/45730 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/497 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/19961 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/46804 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/48081 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/46546 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on Source/WebCore/rendering/RenderTreeAsText.cpp:
```diff
@@ -664,6 +664,7 @@ static void writeLayers(TextStream& ts, const RenderLayer& rootLayer, RenderLaye
         paintDirtyRect.setWidth(std::max<LayoutUnit>(paintDirtyRect.width(), rootLayer.renderBox()->layoutOverflowRect().maxX()));
         paintDirtyRect.setHeight(std::max<LayoutUnit>(paintDirtyRect.height(), rootLayer.renderBox()->layoutOverflowRect().maxY()));
         layer.setSize(layer.size().expandedTo(snappedIntSize(maxLayoutOverflow(layer.renderBox()), LayoutPoint(0, 0))));
+        layer.setNeedsPositionUpdate();
```
🤯

mattwoodrow replied:
This... was a fun one to debug. Doesn't reproduce when loading the testcases in mini browser unsurprisingly.

alanbaradlay commented on Source/WebCore/rendering/RenderObject.cpp:
```diff
@@ -573,8 +573,12 @@ void RenderObject::clearNeedsLayout(HadSkippedLayout hadSkippedLayout)
         renderElement->setLayoutIdentifier(renderElement->view().frameView().layoutContext().layoutIdentifier());
+
+        if (hasLayer())
+            downcast<RenderLayerModelObject>(*this).layer()->setSelfAndChildrenNeedPositionUpdate();
```
It's a bit odd that here at _clear_ we unconditionally mark the layer with selfAndChildrenNeedsPositionUpdate regardless of what actually happened.

mattwoodrow replied:
That's a good point!

I mostly modelled it off the `setLayoutIdentifier` call above, which is a pseudo-dirty-bit for layer tree updates.

If you can think of a better spot, then I'm all ears!

alanbaradlay replied:
> I mostly modelled it off the setLayoutIdentifier call above, which is a pseudo-dirty-bit for layer tree updates.

Not sure how much that code changed, butI initially the simplified layout dirty bit drove the update flow and the layout id was only used to identify which renderer participated in the last layout.

styleWill/DidChange may be a better candidate though I am not sure if certain tree mutation should also trigger setSelfAndChildrenNeedPositionUpdate.

alanbaradlay commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -895,6 +896,37 @@ void RenderLayer::collectLayers(std::unique_ptr<Vector<RenderLayer*>>& positiveZ
+{
+    if (m_layerPositionDirtyBits.containsAny({ LayerPositionUpdates::NeedsPositionUpdate, LayerPositionUpdates::DescendantNeedsPositionUpdate }))
+        return true;
+    if (parent() && parent()->m_layerPositionDirtyBits.contains(LayerPositionUpdates::AllChildrenNeedPositionUpdate))
+        return true;
+    return false;
```
nit. why not just `return parent() && parent()->m_layerPositionDirtyBits.contains(LayerPositionUpdates::AllChildrenNeedPositionUpdate);`

alanbaradlay commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -1132,30 +1307,45 @@ void RenderLayer::recursiveUpdateLayerPositions(RenderElement::LayoutIdentifier
+    }
 
-    if (m_scrollableArea)
+    // FIXME: Verify?
```
Please expand on this.

alanbaradlay reviewed (approved):
Let's see if Simon has some comments on it.

mattwoodrow committed:
Use dirty bits to only traverse the unchanged subtrees with recursiveUpdateLayerPositions.
https://bugs.webkit.org/show_bug.cgi?id=276706
<rdar://131903605>

Reviewed by Alan Baradlay.

Adds a set of dirty bits to mark which RenderLayers need their position updates,
and makes recursiveUpdateLayerPositions only walk the dirty subset of the tree.

With a debug build, the full traversal still gets completed, and it verifies
that all computed values would have been a no-op had they been skipped.

recursiveUpdateLayerPositions gets a templated 'mode' parameter so that the
compiled code in release should be mostly identical to before (except the flag
checks). Each block of state changes handles the write/verify branches, to keep
these together and hopefully keep them in sync.

Also adds a new tree dump to 'Compositing' logging, showing the RenderLayer tree
and its dirty bits before/after the recursiveUpdateLayerPositions.

The general approach is that we mark layers (and their ancestors) that we know
to be changed), and then descend into children of those to check if any ancestor
state that they depend on has changed.

We could instead ensure that changes to state that descendants use trigger a
full descendant subtree update, and avoid needing to iterate the child list to
check. Logging of visited vs updated layer counts suggests that this wouldn't
make much difference. We would likely also need to make sure these extra
triggers are specific (not all style changes for example) to avoid updating too
much.

Layout on a RenderObject with a layer triggers update on the direct children of
that layer, since we can move (but not layout) a descendant renderer, and the
child layer is then a further descendant of the moved renderer. We need to
update the child layer to account for the different offset to parent layer.

ClipRectsContext always (and only) specified
ClipRectsOption::IncludeOverlayScrollbarSize when using RootRelativeClipRects,
so that is now implicit. Clip rect validation now can assert these when adding
to the cache, and verify cached clips using it.

recursiveUpdateLayerPositionsAfterScroll no longer checks isVisuallyNonEmpty()
to clear repaint rects, since this doesn't match the condition we use any other
time (and changing the others results in lots of test failures).

* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::repaintOldAndNewPositionsForSVGRenderer const):
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::ClipRectsCache::ClipRectsCache):
(WebCore::ClipRectsCache::getClipRects const):
(WebCore::RenderLayer::RenderLayer):
(WebCore::RenderLayer::addChild):
(WebCore::RenderLayer::isStackingContextChanged):
(WebCore::RenderLayer::setNeedsPositionUpdate):
(WebCore::RenderLayer::setSelfAndChildrenNeedPositionUpdate):
(WebCore::RenderLayer::setSelfAndDescendantsNeedPositionUpdate):
(WebCore::compositingLogEnabledRenderLayer):
(WebCore::RenderLayer::updateLayerPositionsAfterStyleChange):
(WebCore::RenderLayer::updateLayerPositionsAfterLayout):
(WebCore::RenderLayer::ancestorLayerPositionStateChanged):
(WebCore::RenderLayer::recursiveUpdateLayerPositions):
(WebCore::RenderLayer::dirtyAncestorChainHasSelfPaintingLayerDescendantStatus):
(WebCore::RenderLayer::computeRepaintRects):
(WebCore::RenderLayer::computeRepaintRectsIncludingDescendants):
(WebCore::RenderLayer::compositingStatusChanged):
(WebCore::RenderLayer::recursiveUpdateLayerPositionsAfterScroll):
(WebCore::RenderLayer::dirtyAncestorChainHasBlendingDescendants):
(WebCore::RenderLayer::updateTransform):
(WebCore::RenderLayer::setHasVisibleContent):
(WebCore::RenderLayer::dirtyVisibleContentStatus):
(WebCore::RenderLayer::dirtyAncestorChainVisibleDescendantStatus):
(WebCore::RenderLayer::updateLayerPosition):
(WebCore::RenderLayer::hitTestLayer):
(WebCore::RenderLayer::hitTestTransformedLayerInFragments):
(WebCore::RenderLayer::verifyClipRects):
(WebCore::RenderLayer::verifyClipRect):
(WebCore::RenderLayer::updateClipRects):
(WebCore::RenderLayer::clipRects const):
(WebCore::RenderLayer::calculateClipRects const):
(WebCore::outputLayerPositionTreeLegend):
(WebCore::outputLayerPositionTreeRecursive):
(WebCore::showLayerPositionTree):
* Source/WebCore/rendering/RenderLayer.h:
(WebCore::RenderLayer::setIsHiddenByOverflowTruncation):
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::setBackingSharingLayers):
(WebCore::RenderLayerBacking::setRequiresOwnBackingStore):
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::BackingSharingState::issuePendingRepaints):
(WebCore::RenderLayerCompositor::updateBacking):
* Source/WebCore/rendering/RenderLayerScrollableArea.cpp:
(WebCore::RenderLayerScrollableArea::positionOverflowControls):
* Source/WebCore/rendering/RenderLayerScrollableArea.h:
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::RenderObject::clearNeedsLayout):
(WebCore::operator<<):
* Source/WebCore/rendering/RenderObject.h:

Canonical link: https://commits.webkit.org/284780@main
